state/aws/dynamodb/dynamodb.go

```diff
@@ -127,87 +143,55 @@ func (d *StateStore) BulkGet(req []state.GetRequest) (bool, []state.BulkGetRespo
 
 // Set saves a dynamoDB item.
 func (d *StateStore) Set(req *state.SetRequest) error {
-	value, err := d.marshalToString(req.Value)
+	item, err := d.getItemFromReq(req)
 	if err != nil {
-		return fmt.Errorf("dynamodb error: failed to set key %s: %s", req.Key, err)
+		return err
 	}
 
-	ttl, err := d.parseTTL(req)
-	if err != nil {
-		return fmt.Errorf("dynamodb error: failed to parse ttlInSeconds: %s", err)
+	input := &dynamodb.PutItemInput{
+		Item:      item,
+		TableName: &d.table,
 	}
 
-	var item map[string]*dynamodb.AttributeValue
-	if ttl != nil {
-		item = map[string]*dynamodb.AttributeValue{
-			"key": {
-				S: aws.String(req.Key),
-			},
-			"value": {
-				S: aws.String(value),
-			},
-			d.ttlAttributeName: {
-				N: aws.String(strconv.FormatInt(*ttl, 10)),
-			},
-		}
-	} else {
-		item = map[string]*dynamodb.AttributeValue{
-			"key": {
-				S: aws.String(req.Key),
-			},
-			"value": {
-				S: aws.String(value),
-			},
+	if req.ETag != nil && *req.ETag != "" {
```

Comment:
Since you're doing this, any chance you could support first-write-wins too please? Example from MySQL: https://github.com/dapr/components-contrib/blob/8b48210e3ebab92973a7122f50b2eb1f621a58cd/state/mysql/mysql.go#L453-L478

Reply:
i can take a shot but i would suggest that's beyond the scope of this etag support change. I'll submit a separate pull request for that. though, though this does prompt me to wonder whether dapr ought to require a state store to support ETags in order to support first-write-wins consistency. tracking issue: https://github.com/dapr/components-contrib/issues/1982